### ISSUE TITLE
请问是否支持一个问题中包含多图的微调，是否可以提供构造数据集的相关代码？

### ISSUE THREAD

**Issue by 924973292** [Liuziyu77/Visual-RFT#111]

**Comment by Liuziyu77** [comment]
你可以直接试试以list形式输入图片，看看能不能work，理论上Qwen是支持多图的

**Comment by zfr00** [comment]
请问你解决了嘛

**Comment by 924973292** [comment]
还没有

**Comment by linengcs** [comment]
请问现在有解决么

**Comment by xuexidi** [comment]
> if u want to use 2 images just modify this part, and change the related code in `compute_loss()` of grpo_trained.py to process the multi-image input
> 
>     def make_conversation_image(example):
>         return {
>             "prompt": [
>                 {
>                     "role": "user",
>                     "content": [
>                         {"type": "image"},
>                         {"type": "image"},
>                         {"type": "text", "text": example["problem"]},
>                     ],
>                 },
>             ],
>         }

How to modify the `compute_loss()` to process the multi-image input？

def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        if return_outputs:
            raise ValueError("The GRPOTrainer does not support returning outputs")
    
        

        prompts = [x["prompt"] for x in inputs]
        prompts_text = [maybe_apply_chat_template(example, self.processing_class)["prompt"] for example in inputs]
        images = [x["image"] for x in inputs]
        prompt_inputs = self.processing_class(
            text=prompts_text,
            images=images,
            return_tensors="pt",
            padding=True,
            padding_side="left",
            add_special_tokens=False,
        )
        prompt_inputs = super()._prepare_inputs(prompt_inputs)

        prompt_ids, prompt_mask = prompt_inputs["input_ids"], prompt_inputs["attention_mask"]
        pixel_values = prompt_inputs["pixel_values"]
        image_grid_thw = prompt_inputs["image_grid_thw"]

**Comment by xuexidi** [comment]
> > > if u want to use 2 images just modify this part, and change the related code in `compute_loss()` of grpo_trained.py to process the multi-image input
> > > ```
> > > def make_conversation_image(example):
> > >     return {
> > >         "prompt": [
> > >             {
> > >                 "role": "user",
> > >                 "content": [
> > >                     {"type": "image"},
> > >                     {"type": "image"},
> > >                     {"type": "text", "text": example["problem"]},
> > >                 ],
> > >             },
> > >         ],
> > >     }
> > > ```
> > 
> > 
> > How to modify the `compute_loss()` to process the multi-image input？
> > ```
> > def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
> >         if return_outputs:
> >             raise ValueError("The GRPOTrainer does not support returning outputs")
> >     
> >         
> > 
> >         prompts = [x["prompt"] for x in inputs]
> >         prompts_text = [maybe_apply_chat_template(example, self.processing_class)["prompt"] for example in inputs]
> >         images = [x["image"] for x in inputs]
> >         prompt_inputs = self.processing_class(
> >             text=prompts_text,
> >             images=images,
> >             return_tensors="pt",
> >             padding=True,
> >             padding_side="left",
> >             add_special_tokens=False,
> >         )
> >         prompt_inputs = super()._prepare_inputs(prompt_inputs)
> > 
> >         prompt_ids, prompt_mask = prompt_inputs["input_ids"], prompt_inputs["attention_mask"]
> >         pixel_values = prompt_inputs["pixel_values"]
> >         image_grid_thw = prompt_inputs["image_grid_thw"]
> > ```
> 
> This line only process 1 image by default, so if you pass multi images you should change this to make sure `prompt_inputs=self.processing_class(images=images,....)` work correctly.
> 
> images = [x["image"] for x in inputs]
> And your `inputs[i]["image"]` should have multi image paths like this `{"image":["1.png","2.png",...]}`.

Does self.processing_class support `images=list[list[image]]?`

`prompt_inputs=self.processing_class(images=images,....)`

**Comment by xuexidi** [comment]
> > > > > if u want to use 2 images just modify this part, and change the related code in `compute_loss()` of grpo_trained.py to process the multi-image input
> > > > > ```
> > > > > def make_conversation_image(example):
> > > > >     return {
> > > > >         "prompt": [
> > > > >             {
> > > > >                 "role": "user",
> > > > >                 "content": [
> > > > >                     {"type": "image"},
> > > > >                     {"type": "image"},
> > > > >                     {"type": "text", "text": example["problem"]},
> > > > >                 ],
> > > > >             },
> > > > >         ],
> > > > >     }
> > > > > ```
> > > > 
> > > > 
> > > > How to modify the `compute_loss()` to process the multi-image input？
> > > > ```
> > > > def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
> > > >         if return_outputs:
> > > >             raise ValueError("The GRPOTrainer does not support returning outputs")
> > > >     
> > > >         
> > > > 
> > > >         prompts = [x["prompt"] for x in inputs]
> > > >         prompts_text = [maybe_apply_chat_template(example, self.processing_class)["prompt"] for example in inputs]
> > > >         images = [x["image"] for x in inputs]
> > > >         prompt_inputs = self.processing_class(
> > > >             text=prompts_text,
> > > >             images=images,
> > > >             return_tensors="pt",
> > > >             padding=True,
> > > >             padding_side="left",
> > > >             add_special_tokens=False,
> > > >         )
> > > >         prompt_inputs = super()._prepare_inputs(prompt_inputs)
> > > > 
> > > >         prompt_ids, prompt_mask = prompt_inputs["input_ids"], prompt_inputs["attention_mask"]
> > > >         pixel_values = prompt_inputs["pixel_values"]
> > > >         image_grid_thw = prompt_inputs["image_grid_thw"]
> > > > ```
> > > 
> > > 
> > > This line only process 1 image by default, so if you pass multi images you should change this to make sure `prompt_inputs=self.processing_class(images=images,....)` work correctly.
> > > images = [x["image"] for x in inputs]
> > > And your `inputs[i]["image"]` should have multi image paths like this `{"image":["1.png","2.png",...]}`.
> > 
> > 
> > Does self.processing_class support `images=list[list[image]]?`
> > `prompt_inputs=self.processing_class(images=images,....)`
> 
> yup, Qwen2/2.5 series all support multi-image/videos input, you can check the source code in `site-packages/transformers/model/qwen2_5_vl/processing_qwen2_5_vl.py`
> 
> ![Image](https://github.com/user-attachments/assets/b637e70f-8632-4961-b4d4-c73bc47bd6d1)

Thanks!
From your screenshot, the images param accepts `PIL.Image.Image, np.ndarray, torch.tensor,list[ PIL.Image.Image], list[np.ndarray], list[torch.tensor]`.
Which **do not mentioned** accepts  `images=list[list[image]]`, how did you come to the conclusion that it supports `images=list[list [image]]`?

**Comment by xuexidi** [comment]
> [@xuexidi](https://github.com/xuexidi) yup follow this format list[np.ndarray], actually in the
> 
>         images = [x["image"] for x in inputs]
>         # add the code to read images, like cv.read, or Image.open
>         # then it converts to list[ PIL.Image.Image, list[np.ndarray], list[torch.tensor]
>         prompt_inputs = self.processing_class(
>             text=prompts_text,
>             images=images,
>             return_tensors="pt",
>             padding=True,
>             padding_side="left",
>             add_special_token=False,
>         )

what is the shape of `images`in your code? I want to make it looks like `images=[[image1, image2], [image3], [image4, image5]...]`

**Comment by xuexidi** [comment]
> > > > > if u want to use 2 images just modify this part, and change the related code in `compute_loss()` of grpo_trained.py to process the multi-image input
> > > > > ```
> > > > > def make_conversation_image(example):
> > > > >     return {
> > > > >         "prompt": [
> > > > >             {
> > > > >                 "role": "user",
> > > > >                 "content": [
> > > > >                     {"type": "image"},
> > > > >                     {"type": "image"},
> > > > >                     {"type": "text", "text": example["problem"]},
> > > > >                 ],
> > > > >             },
> > > > >         ],
> > > > >     }
> > > > > ```
> > > > 
> > > > 
> > > > How to modify the `compute_loss()` to process the multi-image input？
> > > > ```
> > > > def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
> > > >         if return_outputs:
> > > >             raise ValueError("The GRPOTrainer does not support returning outputs")
> > > >     
> > > >         
> > > > 
> > > >         prompts = [x["prompt"] for x in inputs]
> > > >         prompts_text = [maybe_apply_chat_template(example, self.processing_class)["prompt"] for example in inputs]
> > > >         images = [x["image"] for x in inputs]
> > > >         prompt_inputs = self.processing_class(
> > > >             text=prompts_text,
> > > >             images=images,
> > > >             return_tensors="pt",
> > > >             padding=True,
> > > >             padding_side="left",
> > > >             add_special_tokens=False,
> > > >         )
> > > >         prompt_inputs = super()._prepare_inputs(prompt_inputs)
> > > > 
> > > >         prompt_ids, prompt_mask = prompt_inputs["input_ids"], prompt_inputs["attention_mask"]
> > > >         pixel_values = prompt_inputs["pixel_values"]
> > > >         image_grid_thw = prompt_inputs["image_grid_thw"]
> > > > ```
> > > 
> > > 
> > > This line only process 1 image by default, so if you pass multi images you should change this to make sure `prompt_inputs=self.processing_class(images=images,....)` work correctly.
> > > images = [x["image"] for x in inputs]
> > > And your `inputs[i]["image"]` should have multi image paths like this `{"image":["1.png","2.png",...]}`.
> > 
> > 
> > Does self.processing_class support `images=list[list[image]]?`
> > `prompt_inputs=self.processing_class(images=images,....)`
> 
> yup, Qwen2/2.5 series all support multi-image/videos input, you can check the source code in `site-packages/transformers/model/qwen2_5_vl/processing_qwen2_5_vl.py`
> 
> ![Image](https://github.com/user-attachments/assets/b637e70f-8632-4961-b4d4-c73bc47bd6d1)

By the way, how to modify the `compute_loss` function if my datasets is a multi-turn chat, like:
```
[
    {"role": "user", "content": [{"type": "text", "text": "text1"}, {"type": "image", "image": "image1.jpg"}]},
    {"role": "assistant", "content": [{"type": "text", "text": "response1"}]},
    {"role": "user", "content": [{"type": "image", "image": "image2.jpg"}]},
    {"role": "assistant", "content": [{"type": "text", "text": "response2"}]},
    {"role": "user", "content": [{"type": "image", "image": "image3.jpg"}]},
    ...
]
```

**Comment by lpw007-maker** [comment]
> > [@xuexidi](https://github.com/xuexidi) yup follow this format list[np.ndarray], actually in the
> > ```
> >     images = [x["image"] for x in inputs]
> >     # add the code to read images, like cv.read, or Image.open
> >     # then it converts to list[ PIL.Image.Image, list[np.ndarray], list[torch.tensor]
> >     prompt_inputs = self.processing_class(
> >         text=prompts_text,
> >         images=images,
> >         return_tensors="pt",
> >         padding=True,
> >         padding_side="left",
> >         add_special_token=False,
> >     )
> > ```
> 
> Hi, I found `images=list[list[images]]` works, thanks!
这里方便贴一下训练集的形式吗....  里面可能单图 可能多图 可能没图  这种应该怎么构建。。
